data-library/library/script/nycoc_checkbook.py

```diff
@@ -0,0 +1,381 @@
+import xml.etree.ElementTree as ET
+import requests
+import pandas as pd
+import datetime as dt
+import calendar
+from pathlib import Path
+import shutil
+import time
+import random
+
+from . import df_to_tempfile
+
+from typing import Literal, TypedDict, Any
+
+api_endpoint = "https://www.checkbooknyc.com/api"
+
+TYPE_OF_DATA = Literal[
+    "Contracts",
+    "Budget",
+    "Revenue",
+    "Payroll",
+    "Spending",
+    "Spending_OGE",
+    "Contracts_OGE",
+    "Spending_NYCHA",
+    "Contracts_NYCHA",
+    "Payroll_NYCHA",
+]
+
+DEFAULT_RECORDS_FROM = "1"
+DEFAULT_MAX_RECORDS = "20000"
+
+DEFAULT_DATE_FORMAT = "%Y-%m-%d"
+DEFAULT_START_PERIOD = dt.date(2010, 1, 1).strftime(DEFAULT_DATE_FORMAT)
+DEFAULT_END_PERIOD = dt.date.today().strftime(DEFAULT_DATE_FORMAT)
+
+DEFAULT_DATA_PATH = Path(__file__).resolve().parent / "capital_spending_data"
+
+USER_AGENTS = [
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 13_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15"
+]
+user_agent = random.choice(USER_AGENTS)
```

Comment:
What's the reasoning for this?

Reply:
Checkbook API starts blocking requests at some point, returning a 500 server error. I found that changing a user-agent param in the request body helps to get more requests through, but it doesn't fully solve the problem.